Fix repair stalls in get_sync_boundary and apply_rows_on_master_in_thread

asias opened the pull request:
This path set fixes stalls in repair that are caused by std::list merge and clear operations during test_latency_read_with_nemesis test. 

Fixes #6940
Fixes #6975
Fixes #6976

avikivity commented on utils/stall_free.hh:
```diff
@@ -0,0 +1,60 @@
+// Similar to std::merge but it does not stall. Must run inside a seastar
+// thread. Merge items from list1 and list2 into list_output. Items from list1
+// can be moved. Items from list2 can only be copied.
+template<class T, class Compare>
```
Need a constraint for Compare

asias replied:
Done.

avikivity commented on utils/stall_free.hh:
```diff
@@ -0,0 +1,60 @@
+// thread. Merge items from list1 and list2 into list_output. Items from list1
+// can be moved. Items from list2 can only be copied.
+template<class T, class Compare>
+std::list<T> stall_free_merge(std::list<T>& list1, const std::list<T>& list2, Compare comp) {
```
list1 should be an rvalue since it is destroyed in the process.

alternatively, call it merge_to(list<T>& to, const list<T>& from); and then you don't have to move from list1 to the result.

asias replied:
I changed to merge_to.

avikivity commented on utils/stall_free.hh:
```diff
@@ -56,5 +56,15 @@ std::list<T> stall_free_merge(std::list<T>& list1, const std::list<T>& list2, Co
+
+// Clear items in a list without stall. Must run inside a seastar thread.
+template<class T>
+void stall_free_clear(std::list<T>& list) {
```
We use the suffix "_gently" in the cache for such things. Let's be consistent and call it merge_gently, clear_gently.

asias replied:
OK.

avikivity commented on repair/row_level.cc:
```diff
@@ -1269,12 +1269,14 @@ class repair_meta {
             // and _working_row_buf are ordered.
             _working_row_buf = utils::stall_free_merge(tmp, row_diff,
                  [this] (const repair_row& x, const repair_row& y) { return _cmp(x.boundary(), y.boundary()) < 0; });
+            utils::stall_free_clear(tmp);
```
With merge_to() you can drop the tmp.swap() call, and then `merge_to(_working_row_bug, row_diff)`. There won't be a tmp that needs clearing.

I think that even with the current code tmp is empty after the merge.

asias replied:
Yes.

avikivity commented on repair/row_level.cc:
```diff
@@ -1269,12 +1269,14 @@ class repair_meta {
                     boost::adaptors::transformed([] (repair_row& r) { thread::maybe_yield(); return r.hash(); }));
         }
         do_apply_rows(row_diff, node_idx, update_buf).get();
+        utils::stall_free_clear(row_diff);
```
It's better to make do_apply_rows() remove elements instead of having a separate loop to clear it afterwards.

asias replied:
Done.

avikivity commented on utils/stall_free.hh:
```diff
@@ -66,5 +66,16 @@ void stall_free_clear(std::list<T>& list) {
 
+template<class T>
+future<> stall_free_clear_outside_thread(std::list<T>& list) {
+    return repeat([&list] () mutable {
```
No need for "outside_thread" if it returns a future.

Also, better to pass by rvalue reference.

asias replied:
I can drop the stall_free_clear version, so we can have a single version that returns a future.

avikivity commented on repair/row_level.cc:
```diff
@@ -1095,24 +1095,32 @@ class repair_meta {
+
+    future<> clear_working_row_buf() {
+        return utils::stall_free_clear_outside_thread(_working_row_buf).then([this] {
+            _working_row_buf_combined_hash.clear();
```
Will this cause a stall too?

asias replied:
No. _working_row_buf_combined_hash is a single hash, not a container.

asias commented:
Changes in 5c3751eb989e4c97133181751aefdfb1ae1e5848:
 - Renamed to _gently
 - Implemented merge_to schematics instead
 - Remove row from the list in do_apply_row when a row is consumed
 - Added constrains for Compare

avikivity commented on configure.py:
```diff
@@ -393,6 +393,7 @@ def find_headers(repodir, excluded_dirs):
     'test/boost/virtual_reader_test',
     'test/boost/bptree_test',
     'test/boost/double_decker_test',
+    'test/boost/stall_free_test',
```
Missing corresponding change to test.py.

asias replied:
Change test.py for what? Do I need to list stall_free_test to test.py? I do not see individual test in test.py is mentioned.

avikivity replied:
Right, that was the old way. The new test.py does not require listing.

asias replied:
Cool.

avikivity commented:
Dequeued, does not build.

```
In file included from test/boost/stall_free_test.cc:23:
./utils/stall_free.hh:58:1: error: ‘future’ does not name a type
   58 | future<> clear_gently(std::list<T>& list) {
      | ^~~~~~
```

Please check with ninja dev-headers, don't just fix the error.

asias commented:
On Tue, Aug 11, 2020 at 6:32 PM Avi Kivity <notifications@github.com> wrote:

> Dequeued, does not build.
>
> In file included from test/boost/stall_free_test.cc:23:
>
> ./utils/stall_free.hh:58:1: error: ‘future’ does not name a type
>
>    58 | future<> clear_gently(std::list<T>& list) {
>
>       | ^~~~~~
>
>
> Please check with ninja dev-headers, don't just fix the error.
>

Fixed the header and checked it does not break ninja dev-headers (there are
pre-existing errors).

New version:

 git@github.com:asias/scylla.git fix_repair_list_stall_merge_clear_v2



—
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/scylladb/scylla/pull/7002#issuecomment-671867123>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AACOETFEPDFM4NZ2K46VTTLSAEM2FANCNFSM4PZVHHXA>
> .
>


-- 
Asias
